[WIP, RFC] Prettify some standard symbols (std::string and containers)

erijo opened the pull request:
This PR is currently a work in progress, but before polishing it I'm interested to hear if you have any comments. Is it a good idea or should I just drop it?

Instead of displaying the full symbol name for some symbols from the std namespace, rewrite them to be more like what you write in the source.

The following is the output from a test program I did:

Prettified symbol | Output from `abi::__cxa_demangle`
--------------------- | ---------------------------------------------
`std::string` | `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >`
`std::wstring` | `std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >`
`std::basic_string<int>` | `std::__cxx11::basic_string<int, std::char_traits<int>, std::allocator<int> >`
`std::vector<int>` | `std::vector<int, std::allocator<int> >`
`std::map<int, float>` | `std::map<int, float, std::less<int>, std::allocator<std::pair<int const, float> > >`
`std::map<std::string, std::vector<std::map<int, std::string>>>` | `std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::map<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<int>, std::allocator<std::pair<int const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::allocator<std::map<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<int>, std::allocator<std::pair<int const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<std::map<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<int>, std::allocator<std::pair<int const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::allocator<std::map<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<int>, std::allocator<std::pair<int const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > > > > >`
`int` | `int`
`TFoo` | `TFoo`
`TBar<std::vector<std::string> >` | `TBar<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >`
`TYou<int, std::string >` | `TYou<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >`
`mystd::TFoo` | `mystd::TFoo`
`mystd::TBar<std::vector<std::string> >` | `mystd::TBar<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >`
`std::string (*)(std::vector<short>)` | `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (*)(std::vector<short, std::allocator<short> >)`
`std::pair<int, int>` | `std::pair<int, int>`
`std::list<std::string>` | `std::__cxx11::list<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >`
`std::set<int>` | `std::set<int, std::less<int>, std::allocator<int> >`
`std::multiset<std::vector<mystd::TFoo>>` | `std::multiset<std::vector<mystd::TFoo, std::allocator<mystd::TFoo> >, std::less<std::vector<mystd::TFoo, std::allocator<mystd::TFoo> > >, std::allocator<std::vector<mystd::TFoo, std::allocator<mystd::TFoo> > > >`
`std::multimap<std::string, std::string>` | `std::multimap<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >`
`std::deque<std::vector<char>>` | `std::deque<std::vector<char, std::allocator<char> >, std::allocator<std::vector<char, std::allocator<char> > > >`
`std::stack<int, std::deque<int> >` | `std::stack<int, std::deque<int, std::allocator<int> > >`
`std::array<int, 3ul>` | `std::array<int, 3ul>`
`std::forward_list<std::list<int>, std::allocator<std::list<int> > >` | `std::forward_list<std::__cxx11::list<int, std::allocator<int> >, std::allocator<std::__cxx11::list<int, std::allocator<int> > > >`
`std::unordered_set<int>` | `std::unordered_set<int, std::hash<int>, std::equal_to<int>, std::allocator<int> >`
`std::unordered_map<int, float>` | `std::unordered_map<int, float, std::hash<int>, std::equal_to<int>, std::allocator<std::pair<int const, float> > >`
`std::unordered_multiset<int>` | `std::unordered_multiset<int, std::hash<int>, std::equal_to<int>, std::allocator<int> >`
`std::unordered_multimap<int, float>` | `std::unordered_multimap<int, float, std::hash<int>, std::equal_to<int>, std::allocator<std::pair<int const, float> > >`

erijo commented:
I've now added a test program to hotspot that can be used to test this.

milianw commented on src/models/data.cpp:
```diff
@@ -133,6 +133,147 @@ ItemCost buildCallerCalleeResult(const BottomUp& data, const Costs& bottomUpCost
     return totalCost;
 }
+
+bool startsWith(const QStringRef& str, const QLatin1String& prefix, int& size)
```
please return a negative size to report a mismatch and don't take the size as an out param

erijo replied:
Would it be acceptable if I changed end to be a pointer? Returning it would make the calling code much uglier.

milianw commented on tests/test-clients/cpp-std-symbols/main.cpp:
```diff
@@ -0,0 +1,61 @@
+template <typename T>
+void doWork(T& t)
+{
+  typename T::value_type value;
```
4 spaces for indentation please

milianw commented on tests/test-clients/cpp-std-symbols/main.cpp:
```diff
@@ -0,0 +1,61 @@
+template <typename T>
+T returnType()
+{
+  T t;
```
here and below too

erijo commented:
> Please write a real unit test that covers this new function. I would like it to go through a list of examples (e.g. all of the ones from your cover letter) and verify that they get prettified as expected

I've added an unit test testing a couple of different variants. Only tested it with libstdc++ (i.e. gcc).

> Also, somewhere we should still show the un-prettified symbol I guess? like inside the tooltips?

Having both the pretty and the orignal symbol at the same place gives a very long string. I've added a menu action to toggle pretty on/off (only implemented for the flame graph, the rest are TDB) as a suggestion. Please let me know what you think.

milianw commented on tests/modeltests/tst_models.cpp:
```diff
@@ -25,6 +25,12 @@
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 
+#include <cxxabi.h>
```
sorry, you misunderstood my intentions. your test is good in coverage but its implementation is overly complicated. Instead, just add a `test_prettifySymbol_data()` and `test_prettifySymbol()`. In the `_data`, add two `QString` columns, one with the "raw" symbol and one with the extected pretty symbol. In the other function, just fetch the two strings, call  prettify on the input and then compare the result to the expected value.
This will also make it trivial to add tests for more complex symbols with high levels of template nesting. Please do test that too, something like:

```
std::__1::__function::__func<std::__1::__bind<bool (foobar::map::api_v2::DeltaAccessImpl::*)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::weak_ptr<stream::Downloader> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&), foobar::map::api_v2::DeltaAccessImpl*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::weak_ptr<stream::Downloader> const&, char const (&) [1], char const (&) [1], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>, std::__1::allocator<std::__1::__bind<bool (foobar::map::api_v2::DeltaAccessImpl::*)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::weak_ptr<stream::Downloader> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&), foobar::map::api_v2::DeltaAccessImpl*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::weak_ptr<stream::Downloader> const&, char const (&) [1], char const (&) [1], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&> >, bool ()
```

Yes, that is a real-world example for a symbol. You can also have a look at the symbols in `nm -aDC tests/test-clients/cpp-parallel/cpp-parallel` for more inspiration of complex symbols. Actually, it would be really cool if you could ensure that the symbols encountered on the hot path in that example are prettified properly:

```
    99.92%     0.00%  cpp-parallel  libc-2.29.so         [.] __GI___clone (inlined)
            |
            ---__GI___clone (inlined)
               |          
                --99.92%--start_thread
                          |          
                           --99.92%--execute_native_thread_routine
                                     std::thread::_State_impl<std::thread::_Invoker<std::tuple<std::__future_base::_Async_state_impl<std::thread::_Invoker<std::tuple<double (*)()> >, doubl>
                                     std::thread::_Invoker<std::tuple<std::__future_base::_Async_state_impl<std::thread::_Invoker<std::tuple<double (*)()> >, double>::_Async_state_impl(std>
                                     std::thread::_Invoker<std::tuple<std::__future_base::_Async_state_impl<std::thread::_Invoker<std::tuple<double (*)()> >, double>::_Async_state_impl(std>
                                     std::__invoke<std::__future_base::_Async_state_impl<std::thread::_Invoker<std::tuple<double (*)()> >, double>::_Async_state_impl(std::thread::_Invoker<>
                                     std::__invoke_impl<void, std::__future_base::_Async_state_impl<std::thread::_Invoker<std::tuple<double (*)()> >, double>::_Async_state_impl(std::thread>
                                     std::__future_base::_Async_state_impl<std::thread::_Invoker<std::tuple<double (*)()> >, double>::_Async_state_impl(std::thread::_Invoker<std::tuple<dou>
                                     std::__future_base::_State_baseV2::_M_set_result (inlined)
                                     std::call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_>
                                     __gthread_once (inlined)
                                     __pthread_once_slow
                                     |          
                                      --99.92%--std::__future_base::_State_baseV2::_M_do_set
                                                std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>::operator() (inlined)
                                                std::_Function_handler<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> (), std::__future_base:>
                                                std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<double>, std::__future_base::_Result_base::_Deleter>, std::thre>
                                                std::thread::_Invoker<std::tuple<double (*)()> >::operator() (inlined)
                                     std::__future_base::_State_baseV2::_M_set_result (inlined)
                                     std::call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_>
                                     __gthread_once (inlined)
                                     __pthread_once_slow
                                     |          
                                      --99.92%--std::__future_base::_State_baseV2::_M_do_set
                                                std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>::operator() (inlined)
                                                std::_Function_handler<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> (), std::__future_base:>
                                                std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<double>, std::__future_base::_Result_base::_Deleter>, std::thre>
                                                std::thread::_Invoker<std::tuple<double (*)()> >::operator() (inlined)
                                                std::thread::_Invoker<std::tuple<double (*)()> >::_M_invoke<0ul> (inlined)
                                                std::__invoke<double (*)()> (inlined)
                                                std::__invoke_impl<double, double (*)()> (inlined)
                                                worker
                                                |          
                                                |--96.22%--std::uniform_real_distribution<double>::operator()<std::linear_congruential_engine<unsigned long, 16807ul, 0ul, 2147483647ul> > (>
                                                |          std::uniform_real_distribution<double>::operator()<std::linear_congruential_engine<unsigned long, 16807ul, 0ul, 2147483647ul> > (>
                                                |          |          
                                                |           --88.87%--std::__detail::_Adaptor<std::linear_congruential_engine<unsigned long, 16807ul, 0ul, 2147483647ul>, double>::operator(>
                                                |                     std::generate_canonical<double, 53ul, std::linear_congruential_engine<unsigned long, 16807ul, 0ul, 2147483647ul> > (in>
                                                |                     |          
                                                |                      --53.43%--std::linear_congruential_engine<unsigned long, 16807ul, 0ul, 2147483647ul>::operator() (inlined)
                                                |                                std::__detail::__mod<unsigned long, 2147483647ul, 16807ul, 0ul> (inlined)
                                                |                                std::__detail::_Mod<unsigned long, 2147483647ul, 16807ul, 0ul, true, true>::__calc (inlined)
                                                |          
                                                 --2.39%--std::norm<double> (inlined)
                                                           std::_Norm_helper<true>::_S_do_it<double> (inlined)
```

The output above was generated with `perf --stdio`, using binutils, so it may be slightly different from what we get from elfutils in hotspot. But I guess you get the idea.

erijo replied:
What c++ code generated the "real world" symbol `std::__1::__function::__func<std::__1::__bind<bool ...`? I'm interested in replicating it.

milianw replied:
that was a symbol reported to me by a colleague from a customer code base we aren't allowed to share.

anyhow, for the purpose of this test, I don't see the need to replicate it using actual C++ code - simplifying operates on strings anyways, so use it as an input and see if it's properly simplified

erijo replied:
I managed to figure it out. It has been added to the unit test.

milianw commented on tests/test-clients/CMakeLists.txt:
```diff
@@ -19,3 +19,4 @@ add_subdirectory(c-syscalls)
 add_subdirectory(cpp-sleep)
 add_subdirectory(cpp-threadnames)
 add_subdirectory(cpp-parallel)
+add_subdirectory(cpp-std-symbols)
```
I don't think this is required or helpful. The string-based testing should be enough.

erijo replied:
Removed.

erijo commented:
I'll have a look at the comments, but do you have any comment on the suggest ui menu toggle option?

milianw commented on src/flamegraph.cpp:
```diff
@@ -195,7 +208,9 @@ QString FrameGraphicsItem::description() const
         totalCost = item->cost();
     }
-    const auto symbol = m_symbol.symbol.isEmpty() ? QObject::tr("??") : m_symbol.prettySymbol;
+    const auto symbol = m_symbol.symbol.isEmpty()
```
wrap this code in a helper function that, given a Data::Symbol, returns either the pretty symbol or not, based on a global settings singleton that's queried

milianw replied:
or ?? if the symbol is invalid

erijo replied:
I did an attempt at this. Please have a look and let me know what you think. Note that BottomUp, TopDown and CallerCallee results pages are not yet updated when the UI setting is toggled. Some pointers on how to correctly invalidate the models there would be helpful.

milianw commented on tests/modeltests/tst_models.cpp:
```diff
@@ -351,28 +406,11 @@ private slots:
     }
 
-    void testSymbol()
+    void testPrettySymbol()
```
please use QTestLib's data-driven tests instead. (cf. QFETCH, QTest::addColumn, QTest::addRow)

but I can clean this up for you ~next week if you don't feel like it

erijo replied:
Ah, now I see what you meant with your previous comment regarding the unit test. Didn't know about QFETCH and friends. I'll fix.

milianw commented on tests/modeltests/tst_models.cpp:
```diff
@@ -161,6 +161,56 @@ const SymbolData prettifySymbolData[] = {
     {"std::unordered_multimap<int, float>",
      "std::unordered_multimap<int, float, std::hash<int>, std::equal_to<int>,"
      " std::allocator<std::pair<int const, float> > >"},
+    {"std::__function::__func<std::__bind<bool (foobar::map::api_v2::DeltaAccessImpl::*)"
```
could / should __function::_func and __bind be prettified too? std::function and std::bind?

milianw replied:
also, please add more coverage for std::future, I think it's not prettified either (see my other comment and check the symbols in the cpp-parallel example)

erijo replied:
My plan was to try to get the basic support in place first and once that has landed work on adding support for more symbols. The ones starting with underscore are a bit special since they are internal symbols that differ between libstdc++ and libc++ so they might require a bit more thought. Is that ok?

erijo replied:
I added a prettification for std::allocator which I think fixes the std::function case.

milianw commented on src/models/data.h:
```diff
@@ -46,23 +46,40 @@ QString prettifySymbol(const QString& symbol);
-    // prettified function name
-    QString prettySymbol;
+    // prettified function name (cached in getSymbol)
+    mutable QString cachedPrettySymbol;
```
considering that most symbols will be prettified by default, I think it would be OK and good to always prettify them on construction of this struct

milianw commented on src/models/data.h:
```diff
@@ -46,23 +46,40 @@ QString prettifySymbol(const QString& symbol);
 
 struct Symbol
 {
+    static bool prettifySymbol;
```
wrong place to store this, introduce a QObject Settings singleton with a getter and setter for this property as well as a notify signal which can then be connected to in the models. The models then need to trigger a dataChanged signal on the respective Name column

milianw commented on src/models/data.h:
```diff
@@ -46,23 +46,40 @@ QString prettifySymbol(const QString& symbol);
     // path to dso / executable
     QString path;
 
+    QString getSymbol() const
```
please remove this logic here and instead add a free function helper to Util that takes a `const Symbol&` input and then checks the settings singleton and returns either the pretty symbol or the un-prettified symbol

milianw commented on src/models/treemodel.cpp:
```diff
@@ -74,7 +74,7 @@ QVariant BottomUpModel::rowData(const Data::BottomUp* row, int column, int role)
         switch (column) {
         case Symbol:
-            return row->symbol.symbol.isEmpty() ? tr("??") : row->symbol.prettySymbol;
+            return row->symbol.getNonEmptySymbol();
```
Util::formatSymbol(row->symbol)

erijo replied:
This and all the other comments should now be fixed. Please have a look.

milianw commented on src/flamegraph.cpp:
```diff
@@ -158,11 +159,12 @@ void FrameGraphicsItem::paint(QPainter* painter, const QStyleOptionGraphicsItem*
-    const auto binary = m_symbol.binary.isEmpty() ? QObject::tr("??") : m_symbol.binary;
-    const auto symbol = m_symbol.symbol.isEmpty() ? QObject::tr("?? [%1]").arg(binary) : m_symbol.symbol;
+    const auto binary = Util::formatString(m_symbol.binary);
+    const auto symbol = Settings::instance()->prettifySymbols() ? m_symbol.prettySymbol : m_symbol.symbol;
```
this should use Util::formatSymbol, no?

milianw commented on src/mainwindow.cpp:
```diff
@@ -160,6 +161,13 @@ MainWindow::MainWindow(QWidget* parent)
     showTimelineAction->setShortcut(tr("Ctrl+T"));
     connect(showTimelineAction, &QAction::toggled, m_resultsPage, &ResultsPage::setTimelineVisible);
 
+    auto *prettifySymbolsAction = ui->viewMenu->addAction(tr("Prettify Symbols"));
```
please also add a tooltip saying something like

"Replace fully qualified and expanded STL type names with their shorter and more commonly used equivalents. E.g. show <tt>std::string</tt> instead of <tt>std::basic_string<char, ...></tt>"

milianw commented on src/mainwindow.cpp:
```diff
@@ -160,6 +161,13 @@ MainWindow::MainWindow(QWidget* parent)
+    auto *prettifySymbolsAction = ui->viewMenu->addAction(tr("Prettify Symbols"));
+    prettifySymbolsAction->setCheckable(true);
+    prettifySymbolsAction->setChecked(Settings::instance()->prettifySymbols());
+    prettifySymbolsAction->setShortcut(tr("Ctrl+Y"));
```
I don't think we need a shortcut for that? If we do, why Ctrl + Y - that's an undo action usually. If you really want a shortcut, please propose an alternative instead.

milianw commented on src/models/callercalleemodel.cpp:
```diff
@@ -88,7 +90,7 @@ QVariant CallerCalleeModel::cell(int column, int role, const Data::Symbol& symbo
         switch (column) {
         case Symbol:
-            return symbol.symbol;
+            return Settings::instance()->prettifySymbols() ? symbol.prettySymbol : symbol.symbol;
```
introduce a Util function for that please

milianw commented on src/models/callercalleemodel.cpp:
```diff
@@ -108,11 +110,13 @@ QVariant CallerCalleeModel::cell(int column, int role, const Data::Symbol& symbo
     } else if (role == FilterRole) {
         // TODO: optimize this
-        return QString(symbol.symbol + symbol.binary);
+        QString result = Settings::instance()->prettifySymbols() ? symbol.prettySymbol : symbol.symbol;
```
Util::formatSymbol

milianw commented on src/models/callercalleemodel.h:
```diff
@@ -83,6 +84,13 @@ class SymbolCostModelImpl : public HashModel<Data::SymbolCostMap, ModelImpl>
+        Parent::connect(Settings::instance(), &Settings::prettifySymbolsChanged, this,
+                        [this]()
+                        {
+                            emit Parent::dataChanged(Parent::index(0, Symbol),
```
return early when the model is empty, i.e.

````
if (!rowCount())
    return;

milianw commented on src/models/callercalleemodel.h:
```diff
@@ -145,7 +153,7 @@ class SymbolCostModelImpl : public HashModel<Data::SymbolCostMap, ModelImpl>
             switch (column) {
             case Symbol:
-                return symbol.symbol;
+                return Settings::instance()->prettifySymbols() ? symbol.prettySymbol : symbol.symbol;
```
Util

milianw commented on src/models/callercalleemodel.h:
```diff
@@ -154,11 +162,13 @@ class SymbolCostModelImpl : public HashModel<Data::SymbolCostMap, ModelImpl>
         } else if (role == FilterRole) {
             // TODO: optimize this
-            return QString(symbol.symbol + symbol.binary);
+            QString result = Settings::instance()->prettifySymbols() ? symbol.prettySymbol : symbol.symbol;
```
Util::formatSymbol

milianw commented on src/models/data.cpp:
```diff
@@ -133,6 +133,158 @@ ItemCost buildCallerCalleeResult(const BottomUp& data, const Costs& bottomUpCost
     return totalCost;
 }
+
+bool startsWith(const QStringRef& str, const QLatin1String& prefix, int* size)
```
static, or into anon namespace, or make it a lambda in the function where it's used

milianw commented on src/models/data.cpp:
```diff
@@ -133,6 +133,158 @@ ItemCost buildCallerCalleeResult(const BottomUp& data, const Costs& bottomUpCost
+    return false;
+}
+
+int findSameDepth(const QStringRef& str, int offset, QChar ch, bool returnNext = false)
```
dito

milianw replied:
either make this a lambda or make it static please

milianw commented on src/models/data.cpp:
```diff
@@ -133,6 +133,158 @@ ItemCost buildCallerCalleeResult(const BottomUp& data, const Costs& bottomUpCost
+            }
+        }
+    }
+    else if (startsWith(symbol, QLatin1String("vector<"), &end)
```
please introduce a lambda that's used here instead of the long conditional chain, and within that lambda use an initializer list over some type names, e.g. something like:

```
const auto types = {QLatin1String("set"), ...};
for (const auto &type : types) {
    if (startsWith(symbol, type, &end))
        return true;
}

milianw commented on src/models/data.cpp:
```diff
@@ -133,6 +133,158 @@ ItemCost buildCallerCalleeResult(const BottomUp& data, const Costs& bottomUpCost
+            symbol = symbol.mid(end);
+        }
+    }
+    else if (startsWith(symbol, QLatin1String("map<"), &end)
```
ah, since this is the same helper function here, let the lambda take the initializer_list of type names as an argument and then use it here again

milianw commented on src/models/data.cpp:
```diff
@@ -133,6 +133,158 @@ ItemCost buildCallerCalleeResult(const BottomUp& data, const Costs& bottomUpCost
+            symbol = symbol.mid(end);
+        }
+    }
+    else if (startsWith(symbol, QLatin1String("allocator<"), &end)) {
```
can you please explain in code comments what the difference between these three conditional branches is? the comment could be as simple as

// translates `set<...>` to `set<...>`

and similar for the three cases, such that it becomes clear what the branches do conceptually without having to grog the code first

milianw commented on src/models/treemodel.h:
```diff
@@ -121,7 +122,10 @@ class TreeModel : public AbstractTreeModel
         if (role == FilterRole) {
             // TODO: optimize
-            return QString(item->symbol.symbol + item->symbol.binary);
+            QString symbol = Settings::instance()->prettifySymbols()
```
Util::formatSymbol

milianw commented on src/settings.h:
```diff
@@ -0,0 +1,50 @@
+public:
+    static Settings* instance();
+
+    Q_PROPERTY(bool prettifySymbols READ prettifySymbols
```
we don't need a Q_PROPERTY, hotspot doesn't use QML or similar

milianw commented on tests/modeltests/tst_models.cpp:
```diff
@@ -323,6 +323,163 @@ private slots:
+            << "std::unordered_multimap<int, float>"
+            << "std::unordered_multimap<int, float, std::hash<int>, std::equal_to<int>,"
+            " std::allocator<std::pair<int const, float> > >";
+        QTest::newRow("bound function")
```
epic :+1:

erijo commented:
I think I've addressed all of your comments.

If you have more just let me know. I'm happy to address to any comments you may have.

milianw commented:
applied to master, thanks a million!

erijo commented:
Closing (merged in e6d04c05fa5).
